### DISCUSSION TITLE
Record selection

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#6460] [Ideas]
We current use `.` for record selection, e.g.:

```
let p = { x = 123, y = 456 };
p.x + p.y
```

But in the future I would like to use `.` for field selection on structs. This means that we need new syntax for record selection. Any ideas? 

(Would it be better to have new syntax for structs?)

**Comment by jaschdoc** [comment]
Structs *could* have `mystruct->field` like C (not sure if I am for it, though)

**Comment by jaschdoc** [comment]
Is it possible to have `.` for both label and field access? Typing should resolve the differences, no?

**Reply by magnus-madsen** [reply]
Probably not because the types are too different that they can be captured by the same type class.

**Comment by stephentetley** [comment]
Maybe use hash?

~~~

let p = { x = 123, y = 456 };
p#x + p#y

~~~

I thought another functional language used hash, but I can only find SML which I think uses it infix. `#x p + #y p`.

**Reply by mlutze** [reply]
I've also seen hash before somewhere. It seems like an OK option.

**Comment by magnus-madsen** [comment]
I have two other proposals.

Given:

```
struct Point {
    x: Int32
    y: Int32
}
```

Proposal 1:

We introduce getters and setters in the companion module, i.e. we automatically derive functions:

```
mod Point {
  def getX(p: Point): Int32
  def setX(v: Int32, p: Point): Unit \ ...
}
```

We can now access the fields using these getters and setters. No new syntax needed.

Proposal 2: Given the above `Point` _inside the companion module_ we treat `e.x` and `e.x = e` as referring to the fields of the struct. This is effectively a (strange) form of shadowing where the meaning of `e.x` now depends on where the expression occurs!

This is a bit crazy-- but it does have one upside: field access and update is automatically restricted to the companion module. 

(Note: This benefit can also be achieved for Proposal 1, if desired).
